tunnel_routes.go

```diff
@@ -0,0 +1,137 @@
+package cloudflare
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// TunnelRoute is the full record for a route.
+type TunnelRoute struct {
+	Network    string    `json:"network"`
+	TunnelId   string    `json:"tunnel_id"`
+	TunnelName string    `json:"tunnel_name"`
+	Comment    string    `json:"comment"`
+	CreatedAt  time.Time `json:"created_at"`
+	DeletedAt  time.Time `json:"deleted_at"`
+}
+
+// tunnelRouteListResponse is the API response for listing tunnel routes.
+type tunnelRouteListResponse struct {
+	Response
+	Result []TunnelRoute `json:"result"`
+}
+
+type tunnelRouteResponse struct {
+	Response
+	Result TunnelRoute `json:"result"`
+}
+
+func extractTunnelRouteResponse(responseBody []byte, err error) (TunnelRoute, error) {
+	if err != nil {
+		return TunnelRoute{}, err
+	}
+
+	var routeResponse tunnelRouteResponse
+	err = json.Unmarshal(responseBody, &routeResponse)
+	if err != nil {
+		return TunnelRoute{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return routeResponse.Result, nil
+}
+
+// TunnelRoutes lists all defined routes for tunnels in the account.
+//
+// See: https://api.cloudflare.com/#tunnel-route-list-tunnel-routes
+func (api *API) TunnelRoutes(ctx context.Context) ([]TunnelRoute, error) {
+	uri := fmt.Sprintf("/%s/%s/teamnet/routes", AccountRouteRoot, api.AccountID)
+	res, err := api.makeRequestContext(ctx, http.MethodGet, uri, nil)
+
+	if err != nil {
+		return []TunnelRoute{}, err
+	}
+
+	var resp tunnelRouteListResponse
+	err = json.Unmarshal(res, &resp)
+	if err != nil {
+		return []TunnelRoute{}, errors.Wrap(err, errUnmarshalError)
+	}
+
+	return resp.Result, nil
+}
+
+// TunnelRouteForIp finds the Tunnel Route that encompasses the given IP
+//
+// See: https://api.cloudflare.com/#tunnel-route-get-tunnel-route-by-ip
+func (api *API) TunnelRouteForIp(ctx context.Context, ip string) (TunnelRoute, error) {
+	uri := fmt.Sprintf("/%s/%s/teamnet/routes/ip/%s", AccountRouteRoot, api.AccountID, ip)
+	return extractTunnelRouteResponse(api.makeRequestContext(ctx, http.MethodGet, uri, nil))
```

Comment:
i don't really like this pattern; you're passing in a full HTTP request without checking the return value and then propagating that to the method here. if the API call fails, this is going to break in all sorts of ways. let's just repeat and handle the errors in the method itself.

Reply:
If you notice the response handling function accepts both the response and and an error, immediately returning if the error is present. I'm happy to change it but I don't think this will fail in the way you're describing.

Reply:
I've inlined this function everywhere it was used